valhalla/sif/dynamiccost.h

```diff
@@ -289,6 +289,52 @@ class DynamicCost {
                               const uint32_t tz_index,
                               uint8_t& restriction_idx) const = 0;
 
+  /**
+   * Checks if edge is bridge, tunnel, toll and if it is
+   * supposed to be excluded.
+   * @param  exclude_flag             Exclude flag value.
+   * @param  pred_class           Predecessor edge classification.
+   * @param  edge_class           Current Edge classification.
+   * @return Returns true if edge is allowed, false if not.
+   */
+  bool IsExcludedRoadClass(const bool exclude_flag,
+                           const baldr::RoadClass pred_class,
+                           const baldr::RoadClass edge_class) const {
+    return exclude_flag && (pred_class != baldr::RoadClass::kMotorway) &&
+           (edge_class == baldr::RoadClass::kMotorway);
+  }
+
+  /**
+   * Checks if any edge exclusion is present.
+   * This checks if bridges, tolls, tunnels, ferries
+   * or highways are excluded in the request.
+   * @param  edge           Pointer to a directed edge.
+   * @param  pred           Predecessor edge information.
+   * @param  exclude_bridges         Flag to exclude bridges.
+   * @param  exclude_tunnels   Flag to exclude tunnels.
+   * @param  exclude_tolls       Flag to exclude tolls.
+   * @param  exclude_highways      Flag to exclude highways
+   * @param  exclude_ferries      Flag to exclude ferries.
+   * @return Returns true if edge should be excluded.
+   */
+
+  bool CheckExclusions(const baldr::DirectedEdge* edge,
+                       const EdgeLabel& pred,
+                       bool exclude_bridges,
+                       bool exclude_tunnels,
+                       bool exclude_tolls,
+                       bool exclude_highways,
+                       bool exclude_ferries) const {
+    return (exclude_bridges && !pred.bridge() && edge->bridge()) ||
+           (exclude_tunnels && !pred.tunnel() && edge->tunnel()) ||
+           (exclude_tolls && !pred.toll() && edge->toll()) ||
+           IsExcludedRoadClass(exclude_highways, pred.classification(), edge->classification()) ||
+           (exclude_ferries &&
+            !(pred.use() == baldr::Use::kFerry || pred.use() == baldr::Use::kRailFerry) &&
+            (edge->use() == baldr::Use::kFerry || edge->use() == baldr::Use::kRailFerry)) ||
+           (edge->is_shortcut() && (exclude_bridges || exclude_tunnels));
```

Comment:
yeah so as i mentioned in my previous comment, this line will be an absolute perf killer. i know we default to false in the config so you cant even use it without opting in however i still wonder if we just remove bridges and tunnels from shortcut building how bad perf in general would drop. my guess is tunnels dont matter but bridges probably really get eaten up by shortcuts. this brings me to another general question why do we want hard avoids of bridges and tunnels? what is the use-case? to me it seems like a poor mans heavy vehicle option, where you arent sure if you'd make it in the height limit of a tunnel or the weight limit of a bridge.

all of that to say i dont want you to make any changes to this. the fact that its opt-in is good enough. i just wanted to ask the question about the use-case.